### PR TITLE
Fix #6765 #6720 improve images

### LINKED ISSUE #6765: Improve Image Samples on the Data Source tab
see [#6720](https://github.com/radiasoft/sirepo/issues/6720)

### PR DESCRIPTION
**Pull request by gurhar1133**

### CODE REVIEW

**Comment by mkeilman** [comment]
I'm currently getting a reshape error with the CIFAR-4 data:
```python
../sirepo/job_driver/__init__.py:279:_agent_receive LocalDriver(a=taLq k=sequential u=2Dzs [_Op(analysis, QR9X)]) error msg={'agentId': 'taLqX1rn6pcfKBuGICu6YPRU9IRmnKn6', 'opId': 'QR9XDULlN7Xf8cvgrn39tpFeridJFSlm', 'opName': 'analysis', 'reply': {'error': 'Invalid shape () for image data', 'stack': 'Traceback (most recent call last):
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/pkcli/job_cmd.py", line 119, in _dispatch_compute
    return _op(expect_file=x)
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/pkcli/job_cmd.py", line 97, in _op
    r = getattr(template_common, f"{msg.jobCmd}_dispatch")(
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/template_common.py", line 738, in stateful_compute_dispatch
    return getattr(t, f"stateful_compute_{m}")(**k, **kwargs)
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 376, in stateful_compute_sample_images
    return _ImagePreview(data).images()
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 1271, in images
    self._append_output_image(y, i)
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 1232, in _append_output_image
    self._gen_image()
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 1217, in _gen_image
    plt.imshow(self.currentImage)
  File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/matplotlib/pyplot.py", line 2695, in imshow
    __ret = gca().imshow(
  File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/matplotlib/__init__.py", line 1446, in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
  File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/matplotlib/axes/_axes.py", line 5663, in imshow
    im.set_data(X)
  File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/matplotlib/image.py", line 710, in set_data
    raise TypeError("Invalid shape {} for image data"
TypeError: Invalid shape () for image data
```

**Comment by gurhar1133** [comment]
> I'm currently getting a reshape error with the CIFAR-4 data:
> 
> ```python
> ../sirepo/job_driver/__init__.py:279:_agent_receive LocalDriver(a=taLq k=sequential u=2Dzs [_Op(analysis, QR9X)]) error msg={'agentId': 'taLqX1rn6pcfKBuGICu6YPRU9IRmnKn6', 'opId': 'QR9XDULlN7Xf8cvgrn39tpFeridJFSlm', 'opName': 'analysis', 'reply': {'error': 'Invalid shape () for image data', 'stack': 'Traceback (most recent call last):
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/pkcli/job_cmd.py", line 119, in _dispatch_compute
>     return _op(expect_file=x)
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/pkcli/job_cmd.py", line 97, in _op
>     r = getattr(template_common, f"{msg.jobCmd}_dispatch")(
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/template_common.py", line 738, in stateful_compute_dispatch
>     return getattr(t, f"stateful_compute_{m}")(**k, **kwargs)
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 376, in stateful_compute_sample_images
>     return _ImagePreview(data).images()
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 1271, in images
>     self._append_output_image(y, i)
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 1232, in _append_output_image
>     self._gen_image()
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 1217, in _gen_image
>     plt.imshow(self.currentImage)
>   File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/matplotlib/pyplot.py", line 2695, in imshow
>     __ret = gca().imshow(
>   File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/matplotlib/__init__.py", line 1446, in inner
>     return func(ax, *map(sanitize_sequence, args), **kwargs)
>   File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/matplotlib/axes/_axes.py", line 5663, in imshow
>     im.set_data(X)
>   File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/matplotlib/image.py", line 710, in set_data
>     raise TypeError("Invalid shape {} for image data"
> TypeError: Invalid shape () for image data
> ```

sorry. I forgot about CIFAR. I'll fix

**Comment by gurhar1133** [comment]
@mkeilman try again?

**Comment by gurhar1133** [comment]
> New error for me:
> 
> ```
> ../sirepo/job_driver/__init__.py:279:_agent_receive LocalDriver(a=LNHj k=sequential u=2Dzs [_Op(analysis, AZer)]) error msg={'agentId': 'LNHjlTj14D3LNe7psqmthiZRVxVWD8iB', 'opId': 'AZermDobNBQ638rsQLV7AkOPbQ6O7Z0Q', 'opName': 'analysis', 'reply': {'error': 'Selection can't process b'bird'', 'stack': 'Traceback (most recent call last):
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/pkcli/job_cmd.py", line 119, in _dispatch_compute
>     return _op(expect_file=x)
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/pkcli/job_cmd.py", line 97, in _op
>     r = getattr(template_common, f"{msg.jobCmd}_dispatch")(
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/template_common.py", line 738, in stateful_compute_dispatch
>     return getattr(t, f"stateful_compute_{m}")(**k, **kwargs)
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 376, in stateful_compute_sample_images
>     return _ImagePreview(data).images()
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 1280, in images
>     self._append_output_image(y, i)
>   File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/activait.py", line 1234, in _append_output_image
>     self.file[self.io.output.label_path][data[index]]
>   File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
>   File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
>   File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/h5py/_hl/dataset.py", line 841, in __getitem__
>     selection = sel.select(self.shape, args, dataset=self)
>   File "/home/vagrant/.pyenv/versions/3.9.15/envs/py3/lib/python3.9/site-packages/h5py/_hl/selections.py", line 82, in select
>     return selector.make_selection(args)
>   File "h5py/_selector.pyx", line 282, in h5py._selector.Selector.make_selection
>   File "h5py/_selector.pyx", line 181, in h5py._selector.Selector.apply_args
> TypeError: Selection can't process b'bird'
> ```
> 
> Probably `pkcompat.from_bytes()` needed somewhere?

I think output should be `metadata/image_types` not `metadata/labels`. Can you try  `metadata/image_types` as the output column? Maybe user needs a better message if wrong column selection doesn't work

**Comment by mkeilman** [comment]
> I think output should be `metadata/image_types` not `metadata/labels`. Can you try `metadata/image_types` as the output column? Maybe user needs a better message if wrong column selection doesn't work

The problem is there's nothing to suggest that any column is somehow "correct" or "incorrect". If we know that ahead of time, then those columns should not be presented or the checkboxes disabled if the user should know the columns exist. If the only way to know is through an error, then yes a meaningful error would help. Do the shapes indicate which columns are compatible? Does it change for input vs output? Maybe some guidance from the experts is in order. I'd relegate all that to a new issue.

File: sirepo/package_data/static/js/activait.js [hunk 1]
```diff
@@ -1096,20 +1096,44 @@ SIREPO.app.directive('imagePreviewPanel', function(requestSender) {
...
           <div data-ng-if="! isLoading() && multiPage">
-            <div class="pull-left">
+            <div data-ng-if="numPages > 3" class="pull-left">
               <button class="btn btn-primary" title="first" data-ng-click="first()">|<</button>
               <button class="btn btn-primary" title="previous" data-ng-disabled="! canUpdateUri(-1)" data-ng-click="prev()"><</button>
             </div>
-            <div class="pull-right">
-                page {{ idx + 1 }} of {{ uris.length }}
+            <div data-ng-if="numPages > 3" class="pull-right">
+                page {{ imageIdx / imagesPerPage + 1 }} of {{ numPages }}
```
Comment (mkeilman) [inline sirepo/package_data/static/js/activait.js:1136]: This page counting is not robust against changes in total number or images or images per page. e.g. change `$scope.imagesPerPage` to 4 ("Page 1 of 3.75"), click to go to the end ("Page 3.75 of 3.75") and then start decrementing - at "Page 0.75 of 3.75" you get a 404 error and a broken image.

File: sirepo/template/activait.py [hunk 2]
```diff
@@ -1228,103 +1218,79 @@ def _x_y(self, run_dir=None):
...
+            self.originals.append(self._pyplot_data_url())
+
+    def _final_images(self):
+        res = PKDict(
+            paramToImage=_param_to_image(self.info),
+            xIsParameters=self.xIsParameters,
+            imageToLabels=self.imageToLabels,
+            pred=False,
+            x=self.inputs,
+            y=self.outputs,
```
Comment (mkeilman) [inline sirepo/template/activait.py:1265]: Can `x` and `y`  be made more descriptive? I'm not sure what they represent, especially on the javascript side where the context here is lost.

**Comment by gurhar1133** [comment]
> > I think output should be `metadata/image_types` not `metadata/labels`. Can you try `metadata/image_types` as the output column? Maybe user needs a better message if wrong column selection doesn't work
> 
> The problem is there's nothing to suggest that any column is somehow "correct" or "incorrect". If we know that ahead of time, then those columns should not be presented or the checkboxes disabled if the user should know the columns exist. If the only way to know is through an error, then yes a meaningful error would help. Do the shapes indicate which columns are compatible? Does it change for input vs output? Maybe some guidance from the experts is in order. I'd relegate all that to a new issue.

Agreed. I made an issue for it [6805](https://github.com/radiasoft/sirepo/issues/6805)

**Review by mkeilman — CHANGES_REQUESTED** [review]
Aside from minor style things, see the comment re image paging

File: sirepo/package_data/static/js/activait.js [hunk 3]
```diff
@@ -1096,20 +1096,44 @@ SIREPO.app.directive('imagePreviewPanel', function(requestSender) {
...
         },
         template: `
-        <div>
-          <img class="img-responsive {{ imageClass }}" />
+        <div class="container-fluid">
+          <div data-ng-if="x" class="row">
+            <div class="col-md-{{ hasThirdColumn ? '4' : '6' }}">
+              <div class="lead text-center">{{ xColName }}</div>
+            </div>
+            <div class="col-md-{{ hasThirdColumn ? '4' : '6' }}">
```
[the remark this answers is not in the document]
Comment (gurhar1133) [reply sirepo/package_data/static/js/activait.js:1106]: Not to my knowledge. We only have three columns if we are doing image to image

File: sirepo/package_data/static/js/activait.js [hunk 4]
```diff
@@ -1096,59 +1096,140 @@ SIREPO.app.directive('imagePreviewPanel', function(requestSender) {
...
+            $scope.predColName = 'Predicted';
+            $scope.imageIdx = 0;
             $scope.dataFileMissing = false;
-            $scope.canUpdateUri = increment => {
-                return $scope.idx + increment >= 0 && $scope.idx + increment < numPages;
+            $scope.hasThirdColumn = true;
+
+            const pageIndex = () => {
+                return $scope.imageIdx / $scope.imagesPerPage;
+            };
```
Comment (mkeilman) [inline sirepo/package_data/static/js/activait.js:1160]: I don't know where we landed re: "braceless" returns (here `const pageIndex = () => $scope.imageIdx / $scope.imagesPerPage;`) - but I advocate consistency within a directive at least. Though `canUpdateUri()` has a complex enough return value that I think it should use braces. Well, I leave it to you

File: sirepo/package_data/static/js/activait.js [hunk 5]
```diff
@@ -1096,59 +1096,140 @@ SIREPO.app.directive('imagePreviewPanel', function(requestSender) {
...
+            function imageInRange(firstImageIndex, rowIndex) {
+                return firstImageIndex + rowIndex + 1 <= $scope.colA.length;
+            }
+
+            function setColumnImage(firstImageIndex, rowIndex, column) {
+                let isTextCol = column == 'A' ? $scope.xIsParams : $scope.imageToLabels;
+                let textCol = column == 'A' ? $scope.parameters : $scope.labels;
+                if (isTextCol) {
+                    if (imageInRange(firstImageIndex, rowIndex)) {
+                        let value = column == 'A' ? `${$scope.colA[firstImageIndex + rowIndex].replace(/[\[\]]/g, '')}` : $scope.colB[firstImageIndex + rowIndex];
```
Comment (mkeilman) [inline sirepo/package_data/static/js/activait.js:1197]: Can be `const`

File: sirepo/package_data/static/js/activait.js [hunk 6]
```diff
@@ -1096,59 +1096,140 @@ SIREPO.app.directive('imagePreviewPanel', function(requestSender) {
...
+            <div data-ng-if="numPages > 1" class="pull-right">
+                page {{ page() }} of {{ numPages }}
               <button class="btn btn-primary" title="next" data-ng-disabled="! canUpdateUri(1)" data-ng-click="next()">></button>
-              <button class="btn btn-primary" title="last" data-ng-click="last()">>|</button>
+              <button class="btn btn-primary" title="last" data-ng-disabled="! canUpdateUri(1)" data-ng-click="last()">>|</button>
             </div>
           </div>
         </div>
         `,
         controller: function($scope, appState) {
```
Comment (mkeilman) [inline sirepo/package_data/static/js/activait.js:1143]: Move `appState` to the directive function

File: sirepo/package_data/static/js/activait.js [hunk 7]
```diff
@@ -1096,59 +1096,140 @@ SIREPO.app.directive('imagePreviewPanel', function(requestSender) {
...
+                        let value = column == 'A' ? `${$scope.colA[firstImageIndex + rowIndex].replace(/[\[\]]/g, '')}` : $scope.colB[firstImageIndex + rowIndex];
+                        textCol.splice(rowIndex, 0, value);
+                        return;
+                    }
+                    textCol.splice(rowIndex, 0, '');
+                    return;
+                }
+                if (imageInRange(firstImageIndex, rowIndex)) {
+                    let value = column == 'A' ? $scope.colA[firstImageIndex + rowIndex] : $scope.colB[firstImageIndex + rowIndex];
+                    $(`.col${column}${$scope.method}${rowIndex + 1}`)[0].src = value;
```
Comment (mkeilman) [inline sirepo/package_data/static/js/activait.js:1206]: Use a constant for the jquery selector. Also, restrict the selector to this element:
```javascript
const colSelector = $($element).find(`.col${column}${$scope.method}${rowIndex + 1}`)[0];
```

File: sirepo/package_data/static/js/activait.js [hunk 8]
```diff
@@ -1096,59 +1096,140 @@ SIREPO.app.directive('imagePreviewPanel', function(requestSender) {
...
                 }
-                if (! $scope.uris) {
+                $(`.col${column}${$scope.method}${rowIndex + 1}`)[0].src = '';
+            }
+
+            function setThirdColumnImage(firstImageIndex, rowIndex) {
+                $scope.hasThirdColumn = $scope.pred != null;
+                if ($(`.pred${$scope.method}1`).length && $scope.hasThirdColumn) {
+                    if (imageInRange(firstImageIndex, rowIndex)) {
+                        $(`.pred${$scope.method}${rowIndex + 1}`)[0].src = $scope.pred[firstImageIndex + rowIndex];
```
Comment (mkeilman) [inline sirepo/package_data/static/js/activait.js:1216]: As above, use a const and $element

File: sirepo/package_data/static/js/activait.js [hunk 9]
```diff
@@ -1096,59 +1096,140 @@ SIREPO.app.directive('imagePreviewPanel', function(requestSender) {
...
-            $scope.uris = null;
-            $scope.idx = 0;
+            $scope.numPages = 0;
+            $scope.imagesPerPage = 3;
+            $scope.pageImages = SIREPO.UTILS.indexArray($scope.imagesPerPage);
+            $scope.colA = null;
+            $scope.colB = null;
+            $scope.pred = null;
+            $scope.ColAName = 'Image';
+            $scope.ColBName = 'Contour';
```
Comment (mkeilman) [inline sirepo/package_data/static/js/activait.js:1152]: `$scope.colAName` etc.

**Review by mkeilman — CHANGES_REQUESTED** [review]
Image paging works great.  Just a handful of items I noticed

**Comment by gurhar1133** [comment]
@moellep ping

**Review by mkeilman — APPROVED** [review]

**Pull request merged**